File: Searching/C++/BinarySearch.cpp
#include <bits/stdc++.h>
using namespace std;

int binarySearch(int* arr, int len, int target){

    int low=0, high=len, mid;
    while(low <= high){
        mid = low + (high-low)/2; // mid = (low+high)/2 
        if (arr[mid] == target){
            return mid; // return if target found at mid
        } else if (arr[mid] > target){
            high = mid-1; 
            // if target is smaller than mid element
            // then it will be present between low & mid-1 
        } else {
            low = mid+1;
            // if target is smaller than mid element
            // then it will be present between mid+1 & high 
        }
    }

    return -1; // to indicate target was not found in array
}

int main(){

    int arr[] = {1, 2, 3, 5, 7, 10, 12, 30, 40, 45};
    // For binary search to work, the array must be sorted
    int target = 40;
    int n = sizeof(arr)/sizeof(arr[0]); // size of array

    int key = binarySearch(arr, n, target);

    if (key == -1){
        cout << "Target element not found" << endl;
    } else {
        cout << "Target element found at position " << key << endl;
    }

    return 0;
}

File: Searching/Java/BinarySearch.java
package Searching.Java;

public class BinarySearch {
    public static void main(String[] args){
        int[] arr = {1,2,4,6,8,9,22,77};
        System.out.println(BinarySearchAlgoritm(arr,22));
    }
    public static int BinarySearchAlgoritm(int[] arr,int target){
        // arr is already sorted array in ascending order;
        int start = 0; 
        int end = arr.length - 1;
        while(start <= end){
            // find middle element in each iteration
            int mid = (start+end)/2; 

            if(target > arr[mid]){ 
                //if target element is greater than the element at the middle than the target must be in right side of mid
                //change the value of start to mid + 1 to search in right part of sorted(ascending) array
                start = mid + 1; 
            } else if(target<arr[mid]){
                //if target element is smaller than the element at mid then the target must lie before the middle element 
                //so we search in the left part of array
                //change the value of end to mid-1;
                end = mid-1;
            } else {
                //target is equal to middle element ie. target == arr[mid]
                //we return the index of middle element which is also equal to the index of target element
                return mid;
            }
        }
        //if we don't find the index of target element we return -1
        return -1;
    }
}
